### DISCUSSION TITLE
Does image subtraction work with both types of Mat?

### DISCUSSION THREAD

**Discussion by Haprst** [bytedeco/javacv#2186] [Q&A] [answered]
I want to subtract one IplImage from another. Since IplImage has no arithmetic methods, I converted them to Mat. But then subtract seems to expect a different type of Mat, and the error message indicates that I should be using opencv.core instead of opencv_core. Is it safe to change just 1 include in 1 file "import org.bytedeco.opencv.opencv_core.Mat"? I have eleven other includes that specify opencv_core in this file, and in 7 other files in this project. The relevant code is the first three lines. Neither of the next two lines works. "No suitable method found for subtract".
[BUG.txt](https://github.com/bytedeco/javacv/files/14304443/BUG.txt)

**Comment by saudet** [comment]
We can convert between both types of Mat using OpenCVFrameConverter when necessary:
http://bytedeco.org/javacv/apidocs/org/bytedeco/javacv/OpenCVFrameConverter.html

**Reply by Haprst** [reply]
That would be great, but it's unclear how to to specify the type of the dst Mat. Is the system smart enough to know that if converting from Mat to Mat, it must mean the other type? Or does the type Mat in line 2 require some prefix?
OpenCVFrameConverter.ToMat convMat = new OpenCVFrameConverter.ToMat();
Mat mat2 = convMat(mat1);

**Reply by saudet** [reply]
There are 2 subclasses, one for each type:
http://bytedeco.org/javacv/apidocs/org/bytedeco/javacv/OpenCVFrameConverter.ToMat.html
http://bytedeco.org/javacv/apidocs/org/bytedeco/javacv/OpenCVFrameConverter.ToOrgOpenCvCoreMat.html

**Reply by Haprst** [reply]
Your answer above was useful, so I've gotten much closer to a solution. But I'm still not there. The relevant code now looks like this, where blueIplImage, greenIplImage, and redIplImage are of course IplImages with matching properties.

There are two sections of the code to test out two different possibilities. In the first section "To Mat" the subtract operation fails at compile time for the reasons shown in my comment. So I commented out that offending line, and then the second section "ToMatMat" fails instead. The first section failed because the subtraction is using Frames instead of Mats. But the second section is unable to convert Frames to Mats. If I use var instead of Mat, the runtime error shows it generated ints instead of Mats.
[TL.txt](https://github.com/bytedeco/javacv/files/14320637/TL.txt)

**Reply by saudet** [reply]
Could you try to call `Loader.load(opencv_java.class)` at the beginning of your program see if that helps?

**Reply by Haprst** [reply]
You're now into the level of systems that's beyond the comprehension of the 84 year-old man. Do you mean that the first statement after  "public static void main(String[] args) {" should be "Loader.load(opencv_java.class);" ? If so, NetBeans immediately pops up the message "cannot find symbol symbol: class opencv_java ... cannot find symbol symbol: Loader"

**Reply by Haprst** [reply]
It then says to add two imports
import org.bytedeco.javacpp.Loader;
import org.bytedeco.opencv.opencv_core.*;
So I did that, followed by a Clean and Build. That compilation gives the same errors as before. No change.

**Reply by Haprst** [reply]
To clarify, it isn't giving any errors associated with the Loader line. It's still giving errors associated with the conversion/confusion between the two types of Mat and Frame.

**Reply by saudet** [reply]
Your IDE should be able to find the packages of those classes, but if not, we can also call with fully-qualified names this way:
`org.bytedeco.javacpp.Loader.load(org.bytedeco.opencv.opencv_java.class)`
As the first statement of the main() method, yes.

**Reply by Haprst** [reply]
To make this easier, I spent a few hours reducing my long 7 file program to a single minimal file. I created a relatively small Test.jpg image which I'm not bothering to send.

At the moment, the 5 problematic statements are undented and surrounded by System.out.println statements, so they're easy to find.  The first two statements are those you suggested earlier today. One is commented out. When the original Loader... statement is active, it requires the first two import statements. When it's replaced with your newer suggestion, those two imports are never used. Neither of those Loader statements solves any of the problems. Is it possible that import statements are causing the problems?
[TLbug.txt](https://github.com/bytedeco/javacv/files/14325624/TLbug.txt)

**Reply by saudet** [reply]
You mean these errors? You'll need to use a `OpenCVFrameConverter.ToMat` to fix those.
```
edeputy.java:77: error: cannot find symbol
    Core.subtract(blueMat1, greenMat1, redMat1);
    ^
  symbol:   variable Core
  location: class edeputy
edeputy.java:99: error: incompatible types: org.opencv.core.Mat cannot be converted to org.bytedeco.opencv.opencv_core.Mat
    Mat blueMat2 = convMatMat1x.convert(blueFrame);
                                       ^
edeputy.java:100: error: incompatible types: org.opencv.core.Mat cannot be converted to org.bytedeco.opencv.opencv_core.Mat
    Mat greenMat2 = convMatMat2x.convert(greenFrame);
                                        ^
edeputy.java:101: error: incompatible types: org.opencv.core.Mat cannot be converted to org.bytedeco.opencv.opencv_core.Mat
    Mat redMat2 = convMatMat3x.convert(redFrame);
                                      ^
edeputy.java:109: error: cannot find symbol
    Core.subtract(blueMat2, greenMat2, redMat2);
    ^
  symbol:   variable Core
  location: class edeputy
5 errors
error: compilation failed

```

**Reply by Haprst** [reply]
Lines 77and 109 are the problem. In both of these lines, the error is Can't find symbol variable Core. For line 77, you say above I need to use  OpenCVFrameConverter.ToMat. But I used exactly that on lines 64 to 66. When that didn't work, and NetBeans didn't suggest a missing include, you suggested ToOrgOpenCvCoreMat. That was the sole reason why I created an alternative 2nd section using instead ToOrgOpenCvCoreMat(). So let's return to line 77. If I used the right method there, then the question is simply why isn't Core resolved? Lines 3 and 4 are imports that reference Core, but evidently something is missing, because Core is not found.

**Reply by Haprst** [reply, marked as answer]
And the answer is that in line 77 "core is not found, but in line 109 "Core" is NOT NEEDED!

**Reply by saudet** [reply]
Right, I don't think you need the Core class, but if you do, it's this one here:
http://bytedeco.org/javacpp-presets/opencv/apidocs/org/opencv/core/Core.html
